### PR TITLE
4390 - New Filing Types: Registrar's Notation, Registrar's Order, Court Order

### PR DESCRIPTION
**Pull request by vysakh-menon-aot**
*Issue #:* /bcgov/entity#4390

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/tests/unit/api/test_business_filings/test_filing_notations.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Copyright © 2019 Province of British Columbia
```
Comment (andrepestana-aot) [inline legal-api/tests/unit/api/test_business_filings/test_filing_notations.py:1]: 2021 :)

**Review by andrepestana-aot — APPROVED** [review]
Looks good to me!

**Comment by vysakh-menon-aot** [comment]
Based on a discussion with Linda and Andre there is a change in validation. So converting this PR to Draft

**Review by andrepestana-aot — APPROVED** [review]
Looks good!

**Comment by vysakh-menon-aot** [comment]
> Are you still changing it?

waiting for schema to be updated. changes are completed

File: legal-api/src/legal_api/services/filings/validations/court_order.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    msg = []
+
+    order_details = get_str(court_order, '/filing/courtOrder/orderDetails')
+
+    if not order_details:
+        msg.append({'error': babel('Court Order is required.'),
+                    'path': '/filing/courtOrder/orderDetails'})
+
+    file_number = get_str(court_order, '/filing/courtOrder/fileNumber')
+    if not file_number:
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/court_order.py:40]: I think this file number required validation is already handled by the court order schema validation?

File: legal-api/src/legal_api/services/authz.py [hunk 3]
```diff
@@ -43,8 +43,10 @@ def authorized(  # pylint: disable=too-many-return-statements
 
     if jwt.has_one_of_roles([BASIC_USER, PUBLIC_USER]):
 
-        # if the action is create_comment, disallow - only staff are allowed
-        if action == 'add_comment':
+        # if the action is create_comment or courtOrder/registrarsNotation/registrarsOrder filings
+        # disallow - only staff are allowed
+        special_actions = ['add_comment', 'court_order', 'registrars_notation', 'registrars_order']
```
Comment (argush3) [inline legal-api/src/legal_api/services/authz.py:48]: Maybe rename to `staff_only_actions`?

**Review by argush3 — APPROVED** [review]
Looks really good.  just two small things I added comments about.

File: legal-api/src/legal_api/services/filings/validations/court_order.py (hunk 2 above)
Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/services/filings/validations/court_order.py:40]: Yes I missed it

File: legal-api/src/legal_api/services/filings/validations/court_order.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from flask_babel import _ as babel  # noqa: N813, I004, I001; importing camelcase '_' as a name
+
+from legal_api.errors import Error
+from legal_api.models import Business
+
+from ...utils import get_str
+# noqa: I003; needed as the linter gets confused from the babel override above.
+
+
+def validate(business: Business, court_order: Dict) -> Error:
```
Comment (thorwolpert) [inline legal-api/src/legal_api/services/filings/validations/court_order.py:27]: return type should be ```Optional[Error]``` as  ln45 ```return None``` is valid

File: legal-api/src/legal_api/services/filings/validations/registrars_notation.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+from flask_babel import _ as babel  # noqa: N813, I004, I001; importing camelcase '_' as a name
+
+from legal_api.errors import Error
+from legal_api.models import Business
+
+from ...utils import get_str
+# noqa: I003; needed as the linter gets confused from the babel override above.
+
+
+def validate(business: Business, registrars_notation: Dict) -> Error:
```
Comment (thorwolpert) [inline legal-api/src/legal_api/services/filings/validations/registrars_notation.py:27]: return type should be ```Optional[Error]``` as  ln46 ```return None``` is valid

File: legal-api/src/legal_api/services/filings/validations/court_order.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    """Validate the Court Order filing."""
+    if not business or not court_order:
+        return Error(HTTPStatus.BAD_REQUEST, [{'error': babel('A valid business and filing are required.')}])
+    msg = []
+
+    order_details = get_str(court_order, '/filing/courtOrder/orderDetails')
```
Comment (thorwolpert) [inline legal-api/src/legal_api/services/filings/validations/court_order.py:33]: consider using the walrus here
```if not (order_details := get_str(court_order, '/filing/courtOrder/orderDetails')):```

**Review by thorwolpert — CHANGES_REQUESTED** [review]
the return type needs updating.

File: legal-api/src/legal_api/services/filings/validations/court_order.py (hunk 6 above)
Comment (thorwolpert) [reply legal-api/src/legal_api/services/filings/validations/court_order.py]: i like you trimmed it down even more!

**Review by thorwolpert — APPROVED** [review]

**Pull request merged**